core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -584,6 +637,23 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
     }
   }
 
+  private def determineNewCatalogPermissions(catalogUpdates: Seq[WorkspaceCatalog], dataAccess: DataAccess, workspaceContext: SlickWorkspaceContext): ReadAction[(Seq[WorkspaceCatalog], Map[Either[RawlsUserRef,RawlsGroupRef], Boolean])] = {
+    for {
+      refsToUpdateByEmail <- dataAccess.rawlsGroupQuery.loadRefsFromEmails(catalogUpdates.map(_.email))
+      existingRefsAndCatalog <- dataAccess.workspaceQuery.findWorkspaceUsersAndCatalog(workspaceContext.workspaceId)
+    } yield {
+      val (emailsFound, emailsNotFound) = catalogUpdates.partition(catalogUpdate => refsToUpdateByEmail.keySet.contains(catalogUpdate.email))
+      val refsToUpdate = catalogUpdates.collect {
+        case x if refsToUpdateByEmail.contains(x.email) => refsToUpdateByEmail(x.email) -> x.catalog
+      }.toSet
```

Comment:
Using `emailsFound.collect` instead of `catalogUpdates.collect` would be slightly more efficient?